### PR TITLE
Add experimental generation and validation of YAML witness invariants with preconditions

### PR DESCRIPTION
**Pull request by sim642**
Adds the experimental YAML witness entry type `precondition_loop_invariant`, which is like `loop_invariant`, but has a second `precondition` expression to match contexts.
Current validation semantics is that the invariant is only checked for contexts where the precondition _must_ hold.

Additionally refactors all the YAML witness type definitions etc to be more type-safe and extensible.

### CODE REVIEW

**Comment by sim642** [comment]
Although the entry type for invariants with preconditions is currently unofficial, it would be good to get this merged for its refactoring of the YAML witnesses. The refactoring makes it easy to work with different entry types, possibly including `location_invariant`, which is to be added as well.

File: src/witness/yamlWitnessType.ml [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+end
+
+module PreconditionLoopInvariantCertificate =
+struct
+  include LoopInvariantCertificate
+  let entry_type = "precondition_loop_invariant_certificate"
+end
+
+(* module type EntryType =
+sig
```
Comment (jerhard) [inline src/witness/yamlWitnessType.ml:273]: Is it intentional to keep this commented out code?

**Review by jerhard — COMMENTED** [review]
We discussed this here  a bit in Munich, with @michael-schwarz, @stilscher.
It probably makes sense to implement this grouping of preconditions before merging this, as otherwise the produced precondition-witnesses could be misleading. 

For a function, one would first generate all the invariant expressions for the contexts, and then check which of the start states may satisfy the condition and group them. Subsequently one would generate the precondition-invariants for these groups of contexts.

File: src/witness/yamlWitness.ml [hunk 2]
```diff
@@ -159,7 +179,42 @@ struct
...
+            offset=Cil.NoOffset;
+            deref_invariant=(fun _ _ _ -> Invariant.none) (* TODO: should throw instead? *)
+          }
+          in
+          begin match Spec.C.invariant context c, Spec.D.invariant context local with
+            | Some c_inv, Some inv ->
+              let loc = Node.location n in
+              (* TODO: group by precondition *)
+              let c_inv = InvariantCil.exp_replace_original_name c_inv in (* cannot be split *)
+              let invs = WitnessUtil.InvariantExp.process_exp inv in
```
Comment (jerhard) [inline src/witness/yamlWitness.ml:200]: Is there a particular advantage to divide conjunctions into separate expressions?

File: src/witness/yamlWitnessType.ml (hunk 1 above)
Comment (sim642) [reply src/witness/yamlWitnessType.ml:273]: Not necessarily. It was an initial attempt to have less duplication and more type safety through a GADT, but for some reason this didn't work out (or at least as nicely).

File: src/witness/yamlWitnessType.ml [hunk 3]
```diff
@@ -0,0 +1,330 @@
+module Producer =
+struct
+  type t = {
+    name: string;
+    version: string;
+    (* TODO: configuration *)
+    command_line: string;
+    (* TODO: description *)
```
Comment (jerhard) [inline src/witness/yamlWitnessType.ml:8]: Are these todos that need to be taken care of before merging this PR?

**Review by jerhard — APPROVED** [review]
Looks good!

File: src/witness/yamlWitnessType.ml (hunk 3 above)
Comment (sim642) [reply src/witness/yamlWitnessType.ml:8]: No, they were also present (just in a different place) before this PR. Those are just optional metadata fields that we currently don't need/use.

**Pull request merged**